core/src/main/scala/tech/beshu/ror/accesscontrol/blocks/rules/DataStreamsRule.scala

```diff
@@ -0,0 +1,92 @@
+import tech.beshu.ror.accesscontrol.blocks.rules.DataStreamsRule.Settings
+import tech.beshu.ror.accesscontrol.blocks.rules.base.Rule
+import tech.beshu.ror.accesscontrol.blocks.rules.base.Rule.RuleResult.{Fulfilled, Rejected}
+import tech.beshu.ror.accesscontrol.blocks.rules.base.Rule.{RegularRule, RuleName, RuleResult}
+import tech.beshu.ror.accesscontrol.blocks.variables.runtime.RuntimeMultiResolvableVariable
+import tech.beshu.ror.accesscontrol.blocks.{BlockContext, BlockContextUpdater}
+import tech.beshu.ror.accesscontrol.domain.DataStreamName
+import tech.beshu.ror.accesscontrol.matchers.ZeroKnowledgeDataStreamsFilterScalaAdapter.CheckResult
+import tech.beshu.ror.accesscontrol.matchers.{MatcherWithWildcardsScalaAdapter, ZeroKnowledgeDataStreamsFilterScalaAdapter}
+import tech.beshu.ror.accesscontrol.request.RequestContext
+import tech.beshu.ror.accesscontrol.utils.RuntimeMultiResolvableVariableOps.resolveAll
+import tech.beshu.ror.utils.ZeroKnowledgeIndexFilter
+
+class DataStreamsRule(val settings: Settings)
+  extends RegularRule {
+
+  override val name: Rule.Name = DataStreamsRule.Name.name
+
+  private val zeroKnowledgeMatchFilter = new ZeroKnowledgeDataStreamsFilterScalaAdapter(
+    new ZeroKnowledgeIndexFilter(true)
+  )
+
+  override def regularCheck[B <: BlockContext : BlockContextUpdater](blockContext: B): Task[Rule.RuleResult[B]] = Task {
+    BlockContextUpdater[B] match {
+      case BlockContextUpdater.DataStreamRequestBlockContextUpdater =>
+        checkDataStreams(blockContext)
+      case _ =>
+        Fulfilled(blockContext)
+    }
+  }
+
+  private def checkDataStreams(blockContext: DataStreamRequestBlockContext): RuleResult[DataStreamRequestBlockContext] = {
+    checkAllowedDataStreams(
+      resolveAll(settings.allowedDataStreams.toNonEmptyList, blockContext).toSet,
+      blockContext.dataStreams,
+      blockContext.requestContext
+    ) match {
+      case Right(filteredDataStreams) => Fulfilled(blockContext.withDataStreams(filteredDataStreams))
+      case Left(()) => Rejected()
+    }
+  }
+
+  private def checkAllowedDataStreams(allowedDataStreams: Set[DataStreamName],
+                                      dataStreamsToCheck: Set[DataStreamName],
+                                      requestContext: RequestContext) = {
+    if (allowedDataStreams.contains(DataStreamName.All) || allowedDataStreams.contains(DataStreamName.Wildcard)) {
+      Right(dataStreamsToCheck)
+    } else {
+      zeroKnowledgeMatchFilter.check(
+        dataStreamsToCheck,
+        MatcherWithWildcardsScalaAdapter.create(allowedDataStreams)
+      ) match {
+        case CheckResult.Ok(processedDataStreams) if requestContext.isReadOnlyRequest =>
+          Right(processedDataStreams)
+        case CheckResult.Ok(processedDataStreams) if processedDataStreams.size === dataStreamsToCheck.size =>
+          Right(processedDataStreams)
+        case CheckResult.Ok(_) | CheckResult.Failed =>
```

Comment:
let's split these two cases and add debug logs with an explanation of what has just happened (please do the same in case of RepositoriesRule). 

We often explain to users why some rules are mismatched. They should be able to say it using debug logs analysis.